tensorflow/core/kernels/unique_op_gpu.cu.cc

```diff
@@ -0,0 +1,443 @@
+                    U** tensor_data, DoneCallback done) const {
+    OP_REQUIRES_OK_ASYNC(context,
+                         context->allocate_temp(DataTypeToEnum<U>::value,
+                                                TensorShape({size}), tensor),
+                         done);
+    *tensor_data = tensor->flat<U>().data();
+  }
+
+  void ComputeAsync(OpKernelContext* context, DoneCallback done) override {
+    const Tensor& input = context->input(0);
+    // TODO(dga):  Make unique polymorphic for returning int32 and int64
+    // vectors to support large tensors.
+    OP_REQUIRES_ASYNC(context,
+                      input.NumElements() <= std::numeric_limits<int32>::max(),
+                      errors::InvalidArgument(
+                          "unique does not support input tensors larger than ",
+                          std::numeric_limits<int32>::max(), " elements"),
+                      done);
+
+    OP_REQUIRES_ASYNC(context, TensorShapeUtils::IsVector(input.shape()),
+                      errors::InvalidArgument("unique expects a 1D vector."),
+                      done);
+
+    se::Stream* stream = context->op_device_context()->stream();
+    OP_REQUIRES_ASYNC(context, stream,
+                      errors::Internal("No GPU stream available."), done);
+
+    int64 input_size = input.NumElements();
+
+    // The algorithm implemented here is as follows:
+    // 1) Sort the input to group equal values together in segments.
+    //      sorted_input, sorted_input_inds = sort(input)
+    // 2) Identify the boundaries between segments and use prefix sum to
+    //    compute the unique ID for each sorted value.
+    //      sorted_input_unique_ids = prefix_sum(indicator(sorted_input))
+    // 3) Extract the input index of the first occurrence of each unique value.
+    //    If counts are required, also extract the end index of each segment.
+    //      unique_input_inds[sorted_input_unique_ids] =
+    //          sorted_input_inds (@ indicator)
+    //      segment_ends[sorted_input_unique_ids[i] - 1] = i (@ indicator)
+    // 4) Sort the extracted unique input indices to put them in order of
+    //    first appearance.
+    //      sorted_unique_input_inds, sorted_unique_perm =
+    //          sort(unique_input_inds)
+    // 5) Gather the sorted unique input values to produce output, and invert
+    //    the second sort permutation to produce an inverse ID mapping. If
+    //    counts are required, also take the adjacent difference between
+    //    segment_ends indices to produce counts.
+    //      output = input[sorted_unique_input_inds]
+    //      inv_sorted_unique_perm[sorted_unique_perm[i]] = i
+    //      counts = adjacent_difference(segment_ends)
+    // 6) Look up unique IDs via the inverse ID mapping and scatter them using
+    //    the original sort permutation to produce the indices output.
+    //      idx[sorted_input_inds] =
+    //          inv_sorted_unique_perm[sorted_input_unique_ids]
```

Comment:
Would it be too cumbersome to interleaved a small but non-trivial running example with the algorithm?  That would make the algorithm much easier to understand at a glance.

Reply:
Done.